### ISSUE TITLE
[analyzer] security.ArrayBound false positive: the same virtual getter is assumed to return two different values on the same path

### ISSUE THREAD

**Issue by earnol** [llvm/llvm-project#222699]
## Summary
When a virtual member function cannot be devirtualized, each call site is evaluated conservatively and gets its own conjured return symbol. Nothing relates the symbols, so the engine can assume contradictory values for two reads of the same getter on the same object within a single path. `security.ArrayBound` then compares an offset against an extent that path could never have, and reports an out-of-bounds write that cannot occur.

## Reproducer
```c++
struct Msg
{
  virtual unsigned cmd() const = 0;
};

void dispatch(Msg* m)
{
  if (m->cmd() != 12)
    return;
  unsigned* d = new unsigned[m->cmd() == 12 ? 2 : 1];
  d[1] = 0;
  delete[] d;
}
```

```sh
clang++ --analyze -Xclang -analyzer-checker=core,security.ArrayBound -Xclang -analyzer-output=text repro.cpp
```

Godbolt (clang trunk): [example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:24,endLineNumber:8,positionColumn:24,positionLineNumber:8,selectionStartColumn:24,selectionStartLineNumber:8,startColumn:24,startLineNumber:8),source:'+struct+Msg%0A++%7B%0A++++virtual+unsigned+cmd()+const+%3D+0%3B%0A++%7D%3B%0A%0A++void+dispatch(Msg*+m)%0A++%7B%0A++++if+(m-%3Ecmd()+!!%3D+12)%0A++++++return%3B%0A++++unsigned*+d+%3D+new+unsigned%5Bm-%3Ecmd()+%3D%3D+12+%3F+2+:+1%5D%3B%0A++++d%5B1%5D+%3D+0%3B%0A++++delete%5B%5D+d%3B%0A++%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:54.412532637075714,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'1',binary:'1',binaryObject:'0',commentOnly:'1',debugCalls:'1',demangle:'1',directives:'1',execute:'1',intel:'1',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'--analyze+-Xclang+-analyzer-checker%3Dcore,security.ArrayBound+-Xclang+-analyzer-output%3Dtext',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'x86-64+clang+3.4+(assertions)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),k:45.587467362924286,l:'4',m:100,n:'0',o:'',s:1,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

## Actual
```
repro.cpp:11:3: warning: Out of bound access to memory after the end of the heap area [security.ArrayBound]
repro.cpp:8:7: note: Assuming the condition is false
repro.cpp:8:3: note: Taking false branch
repro.cpp:10:30: note: Assuming the condition is false
repro.cpp:10:30: note: '?' condition is false
repro.cpp:11:3: note: Access of the heap area at index 1, while it holds only a single 'unsigned int' element
1 warning generated.
```

## Expected
No diagnostic.
The two notes contradict each other. At line 8 `m->cmd() != 12` is assumed false, so `m->cmd()` is 12 on this path. At line 10 `m->cmd() == 12` is assumed false for the same call on the same object, with no write in between. The array therefore has 2 elements on every real execution and `d[1] = 0;` is in bounds.

## Workaround
Making the getter non-virtual, with everything else unchanged, produces no diagnostic:
```c++
struct Msg
{
  unsigned v;
  unsigned cmd() const { return v; }   // non-virtual
};
```

Keeping it virtual but moving the body into the same translation unit still reports, so the missing definition is not the trigger because the failure to devirtualize is.

## Notes
- If the analyzer can devirtualize the receiver, the report disappears. Putting the driver `main()` above in the same translation unit is enough to suppress it, which is why the driver is kept separate.
- `-analyzer-config crosscheck-with-z3=true` does not filter the report, checked on a build with `LLVM_ENABLE_Z3_SOLVER=ON`. The collected constraints are satisfiable, since the two symbols are distinct as far as the solver is concerned.

**Comment by NagyDonat** [comment]
I see why is this report unwanted, but I still say that **this is a true positive report**.

An opaque method (e.g. one that cannot be devirtualized, or one that's simply defined in another TU) can do anything and analyzer has no reason to assume that calling `cmd()` twice on the same object returns the same value twice: it could just as well return a fresh random number, the value of a counter which is incremented after each call, the result of executing a system command etc.

There is no established way to mark that a method is a "getter" and will return the same value when called on the same object (even a name starting with `get` is not a guarantee: e.g. the analyzer source code is full of `getWhatever` methods that are not pure getters).

If we would work with codebases where the getter methods (especially more complex, e.g. virtual ones) have annotations that mark them as pure getters, then it would be reasonable to develop a checker that would model that calling a getter method on the same object always produces the same result. Without that, I don't think that there is anything to do.

**Comment by earnol** [comment]
I can provide a [counter example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:32,positionColumn:1,positionLineNumber:32,selectionStartColumn:1,selectionStartLineNumber:32,startColumn:1,startLineNumber:32),source:'struct+Msg%0A%7B%0A++virtual+~Msg()+%7B%7D%0A%0A++virtual+unsigned+cmd()+const+%3D+0%3B%0A%7D%3B%0A%0Astruct+Ctrl+:+Msg%0A%7B%0A++unsigned+c%3B%0A%0A++explicit+Ctrl(unsigned+cmd)+:+c(cmd)+%7B%7D%0A%0A++unsigned+cmd()+const+final+%7B+return+c%3B+%7D%0A%7D%3B%0A%0Astatic+unsigned+slots(const+Ctrl*+m)%0A%7B%0A++return+m-%3Ecmd()+%3D%3D+12+%3F+2+:+1%3B%0A%7D%0A%0Avoid+dispatch(Ctrl*+c)%0A%7B%0A++if+(c-%3Ecmd()+!!%3D+12)%0A++++return%3B%0A%0A++unsigned*+d+%3D+new+unsigned%5Bslots(c)%5D%3B%0A++d%5B1%5D+%3D+0%3B++//+reported+here%0A++delete%5B%5D+d%3B%0A%7D%0A%0Astatic+Ctrl+incoming(12)%3B%0Astatic+void+(*handler)(Ctrl*)+%3D+dispatch%3B%0A%0Aint+main()%0A%7B%0A++handler(%26incoming)%3B%0A++return+0%3B%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:54.412532637075714,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'1',binary:'1',binaryObject:'0',commentOnly:'1',debugCalls:'1',demangle:'1',directives:'1',execute:'1',intel:'1',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'--analyze+-Xclang+-analyzer-checker%3Dcore,security.ArrayBound+-Xclang+-analyzer-output%3Dtext',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'x86-64+clang+3.4+(assertions)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),k:45.587467362924286,l:'4',m:100,n:'0',o:'',s:1,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) (less reduced method) where hidden method is available and I'm still getting the report.

I'm not claiming it's being minimal but it's quite small:
```c++
struct Msg
{
  virtual ~Msg() {}
  virtual unsigned cmd() const = 0;
};

struct Ctrl : Msg
{
  unsigned c;
  explicit Ctrl(unsigned cmd) : c(cmd) {}
  unsigned cmd() const final { return c; }
};

static unsigned slots(const Ctrl* m)
{
  return m->cmd() == 12 ? 2 : 1;
}

void dispatch(Ctrl* c)
{
  if (c->cmd() != 12)
    return;

  unsigned* d = new unsigned[slots(c)];
  d[1] = 0;  // reported here
  delete[] d;
}

static Ctrl incoming(12);
static void (*handler)(Ctrl*) = dispatch;

int main()
{
  handler(&incoming);
  return 0;
}
```

> An opaque method (e.g. one that cannot be devirtualized, or one that's simply defined in another TU) can do anything and analyzer has no reason to assume that calling `cmd()` twice on the same object returns the same value twice: it could just as well return a fresh random number, the value of a counter which is incremented after each call, the result of executing a system command etc.

I fully agree with this statement, however analyzer has no reason to assume that calling `cmd()` twice on the same object returns the same value twice, but no reason it returns different values either. My reasons are: 
- `cmd()` is `const`
- class `Msg` has no mutable members whatsoever


So my point is we have zero pointers to the fact that `cmd()` might not be a pure getter. And under this light why analyzer assumes that it's a not a pure getter? 
Please note we are getting no detect in this code:
```c++
extern unsigned cmd;  // defined in another translation unit

void dispatch()
{
  if (cmd != 12)
    return;

  unsigned* d = new unsigned[cmd == 12 ? 2 : 1];
  d[1] = 0;
  delete[] d;
}
```
which is completely expected because the value of `cmd` is unknown. So why analyzer suddenly behave differently between those two cases?

**Comment by NagyDonat** [comment]
> I can provide a [counter example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:32,positionColumn:1,positionLineNumber:32,selectionStartColumn:1,selectionStartLineNumber:32,startColumn:1,startLineNumber:32),source:'struct+Msg%0A%7B%0A++virtual+~Msg()+%7B%7D%0A%0A++virtual+unsigned+cmd()+const+%3D+0%3B%0A%7D%3B%0A%0Astruct+Ctrl+:+Msg%0A%7B%0A++unsigned+c%3B%0A%0A++explicit+Ctrl(unsigned+cmd)+:+c(cmd)+%7B%7D%0A%0A++unsigned+cmd()+const+final+%7B+return+c%3B+%7D%0A%7D%3B%0A%0Astatic+unsigned+slots(const+Ctrl*+m)%0A%7B%0A++return+m-%3Ecmd()+%3D%3D+12+%3F+2+:+1%3B%0A%7D%0A%0Avoid+dispatch(Ctrl*+c)%0A%7B%0A++if+(c-%3Ecmd()+!!%3D+12)%0A++++return%3B%0A%0A++unsigned*+d+%3D+new+unsigned%5Bslots(c)%5D%3B%0A++d%5B1%5D+%3D+0%3B++//+reported+here%0A++delete%5B%5D+d%3B%0A%7D%0A%0Astatic+Ctrl+incoming(12)%3B%0Astatic+void+(*handler)(Ctrl*)+%3D+dispatch%3B%0A%0Aint+main()%0A%7B%0A++handler(%26incoming)%3B%0A++return+0%3B%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:54.412532637075714,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'1',binary:'1',binaryObject:'0',commentOnly:'1',debugCalls:'1',demangle:'1',directives:'1',execute:'1',intel:'1',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'--analyze+-Xclang+-analyzer-checker%3Dcore,security.ArrayBound+-Xclang+-analyzer-output%3Dtext',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'x86-64+clang+3.4+(assertions)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),k:45.587467362924286,l:'4',m:100,n:'0',o:'',s:1,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) (less reduced method) where hidden method is available and I'm still getting the report.

This concrete example **IS** a false positive, the analyzer should be able to devirtualize and inline the call of `cmd()` which is declared `final` and called on a pointer with static type `Ctrl *`.

You should probably open a separate issue for it (with a more minified example that omits `dispatch()` and just demonstrates the fact that `cmd()` fails to be inlined in these circumstances -- perhaps with the help of `debug` checkers). AFAIK the devirtualization code is a bit shaky (although I'm not familiar with it), so it wouldn't be surprising if it contains a bug like this.

--------

However, when the analyzer is unable to devirtualize the call and inline the definition, the analyzer has no reason to magically realize that "this must be a getter method".

> I fully agree with this statement, however has no reason to assume that calling `cmd()` twice on the same object returns the same value twice returns different values either. My reasons are:
> 
>     * `cmd()` is `const`
> 
>     * class `Msg` has no mutable members whatsoever
> 
> 
> So my point is we have zero pointers to the fact that `cmd()` might not be a pure getter. And under this light why analyzer assumes that it's a not a pure getter?

Even if the method cannot change the state _of the instance_, it can change the state of many other things (global variables, singletons, objects to which it has a pointer, the network/user/etc.) and those changes can lead to returning a different value in the second call.

For example if you tweak your code to
```c++
struct Msg
{
  virtual ~Msg() {}
  virtual unsigned cmd() const = 0;
};

struct Ctrl : Msg
{
  unsigned *c;
  explicit Ctrl(unsigned *p) : c(*p) {}
  unsigned cmd() const final { return (*c)++; }
};
```
the method is still `const`, there are still no mutable data members, but each call of `cmd()` returns a different number.

Even if the class has no pointer members or no members at all, it can still e.g. return user input or change globals (or singletons which are just fancy globals) and return different values based on such changes.

> [after the example with `extern unsigned cmd;`] So why analyzer suddenly behave differently between those two cases?

When `cmd` is an `extern unsigned cmd` value, referencing it just loads a value from the memory. When `cmd()` is a function (and the analzyer does not resolve it), calling it can execute arbitrary code that mutate and observe the globals and the external world – so there is no reason to assume that it produces the same result. This is not Haskell where the functions are pure, mathematical functions – when you call a C/C++ function, it can do everything, up to launching the nukes (as the Haskellers tend to joke).
